File: lib/classes.js
'use strict';

const TR_CLASSES = {};
const CLASSES = require('../data/classes').reduce((a, c) => {
	const vanilla = c.vanilla.toLowerCase();
	a[vanilla] = c.data;
	if(vanilla !== 'miner') {
		TR_CLASSES[c.data.toLowerCase()] = c.vanilla;
	}
	return a;
}, {});

function getReplacementClass(c, mode) {
	if(mode === 'PT') {
		return CLASSES[c?.toLowerCase()];
	} else if(mode === 'TR') {
		return TR_CLASSES[c?.toLowerCase()];
	}
}

module.exports = {
	onRecord(record, mode) {
		if(record.type === 'Npc') {
			const replacementClass = getReplacementClass(record.class, mode);
			if(replacementClass) {
				console.log(record.type, record.id, 'has class', record.class, 'which should be', replacementClass);
			}
		}
	},
	onInfo(record, topic, mode) {
		if(record.speaker_class) {
			const replacementClass = getReplacementClass(record.speaker_class, mode);
			if(replacementClass) {
				console.log(record.type, record.info_id, 'in topic', topic.id, 'has a', record.speaker_class, 'filter');
			}
		}
		record.filters?.forEach(filter => {
			if(filter.filter_type === 'NotClass') {
				const replacementClass = getReplacementClass(filter.id, mode);
				if(replacementClass) {
					console.log(record.type, record.info_id, 'in topic', topic.id, 'has a Not Class', filter.id, 'filter');
				}
			}
		});
	}
};

File: src/validators/classes.rs
use std::collections::HashMap;

use super::Context;
use crate::{context::Mode, handlers::Handler};
use tes3::esp::{Dialogue, DialogueInfo, FilterType, TES3Object};

include!(concat!(env!("OUT_DIR"), "/gen_classes.rs"));

pub struct ClassValidator {
    tr_classes: HashMap<&'static str, &'static str>,
    classes: HashMap<&'static str, &'static str>,
}

impl Handler<'_> for ClassValidator {
    fn on_record(&mut self, context: &Context, record: &TES3Object) {
        if let TES3Object::Npc(npc) = record {
            if !npc.class.is_empty() {
                if let Some(replacement) = self.get_replacement(&npc.class, context) {
                    println!(
                        "Npc {} has class {} which should be {}",
                        &npc.id, &npc.class, replacement
                    );
                }
            }
        }
    }

    fn on_info(&mut self, context: &Context, record: &DialogueInfo, topic: &Dialogue) {
        if !record.speaker_class.is_empty()
            && self
                .get_replacement(&record.speaker_class, context)
                .is_some()
        {
            println!(
                "Info {} in topic {} has a {} filter",
                record.id, topic.id, record.speaker_class
            )
        }
        for filter in &record.filters {
            if filter.filter_type == FilterType::NotClass
                && self.get_replacement(&filter.id, context).is_some()
            {
                println!(
                    "Info {} in topic {} has a Not Class {} filter",
                    record.id, topic.id, filter.id
                );
            }
        }
    }
}

impl ClassValidator {
    pub fn new() -> Self {
        let (tr_classes, classes) = get_class_data();
        Self {
            tr_classes,
            classes,
        }
    }

    fn get_replacement(&self, id: &str, context: &Context) -> Option<&&'static str> {
        if context.mode == Mode::PT {
            return self.classes.get(id.to_ascii_lowercase().as_str());
        } else if context.mode == Mode::TR {
            return self.tr_classes.get(id.to_ascii_lowercase().as_str());
        }
        None
    }
}
